clusterloader2/pkg/measurement/common/api_availability_measurement.go

```diff
@@ -0,0 +1,209 @@
+	for _, host := range a.hosts {
+		// SSH and check the health of the host
+		command := fmt.Sprintf("curl -f -s -k %slocalhost:%v/healthz", "https://", 443)
+		sshResult, err := measurementutil.SSH(command, host+":22", provider)
+		if err != nil || sshResult.Code != 0 {
+			a.updateAvailabilityMetrics(false, a.hostLevelMetrics[host])
+		} else {
+			a.updateAvailabilityMetrics(true, a.hostLevelMetrics[host])
+		}
+	}
+}
+
+func (a *apiAvailabilityMeasurement) updateClusterAvailabilityMetrics(c clientset.Interface) {
+	// Check the availability of the cluster by issuing a REST call to /healthz end point
+	result := c.CoreV1().RESTClient().Get().AbsPath("/healthz").Do(context.TODO())
+	status := 0
+	result.StatusCode(&status)
+	if status == 200 {
+		a.updateAvailabilityMetrics(true, a.clusterLevelMetrics)
+	} else {
+		a.updateAvailabilityMetrics(false, a.clusterLevelMetrics)
+	}
+}
+
+func (a *apiAvailabilityMeasurement) start(config *measurement.Config, SSHToMasterSupported bool, probeDuration int) error {
+	a.hosts = config.ClusterFramework.GetClusterConfig().MasterIPs
+	if len(a.hosts) < 1 {
+		return goerrors.Errorf("API Server measurement can't execute due to no MasterIPs")
+	}
+
+	k8sClient := config.ClusterFramework.GetClientSets().GetClient()
+	provider, err := util.GetStringOrDefault(config.Params, "provider", config.ClusterFramework.GetClusterConfig().Provider)
+	if err != nil {
+		return err
+	}
+
+	a.isRunning = true
+	a.stopCh = make(chan struct{})
+	a.wg.Add(1)
+
+	go func() {
+		defer a.wg.Done()
+		for {
+			select {
+			case <-a.stopCh:
+				return
+			case <-time.After(time.Duration(probeDuration)):
+				a.updateClusterAvailabilityMetrics(k8sClient)
+				if SSHToMasterSupported {
+					a.updateMasterAvailabilityMetrics(k8sClient, config, provider)
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+// Execute starts the api-server healthz probe end point from start action and
+// collects availability metrics in gather.
+func (a *apiAvailabilityMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	SSHToMasterSupported := config.ClusterFramework.GetClusterConfig().SSHToMasterSupported
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+
+	probeFrequency, err := util.GetIntOrDefault(config.Params, "frequency", 1)
+	if err != nil {
+		return nil, err
+	}
+
+	switch action {
+	case "start":
+		if a.isRunning {
+			klog.Infof("%s: measurement already running", a)
+			return nil, nil
+		}
+		return nil, a.start(config, SSHToMasterSupported, probeFrequency)
+	case "gather":
+		err := a.stop(SSHToMasterSupported, probeFrequency)
+		if err != nil {
+			return nil, err
+		}
+		return a.summaries, nil
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+func (a *apiAvailabilityMeasurement) createClusterAvailabilitySummary(probeFrequency int) error {
+	a.clusterLevelMetrics.updateMaxConsecutiveFailuresIfNeeded()
```

Comment:
Instead of doing this here, let's:
- remove it from here (and createMasterAvailabilitySummary)
- let's explicitly add it `updateFailureMetrics`, i.e.;

```
func (a *apiAvailabilityMetrics) updateFailureMetrics() {
	a.numFailures++
	a.consecutiveFailedProbes++
        updateMaxConsecutiveFailuresIfNeeded()
}
```

With that - you don't need to call it anywhere elese.

Reply:
Done. That makes sense.